Document/0x05d-Testing-Data-Storage.md

```diff
@@ -310,6 +406,36 @@ Inspect the source code to determine whether native Android mechanisms identify
 - Make sure that the app is using the Android KeyStore and Cipher mechanisms to securely store encrypted information on the device. Look for the patterns `AndroidKeystore`, `import java.security.KeyStore`, `import javax.crypto.Cipher`, `import java.security.SecureRandom`, and corresponding usages.
 - Use the `store(OutputStream stream, char[] password)` function to store the KeyStore to disk with a password. Make sure that the password is provided by the user, not hard-coded.
 
+##### Storing a Key - example
+
+To mitigate unauthorized use of keys on the Android device, Android KeyStore lets apps specify authorized uses of their keys when generating or importing the keys. Once made, authorizations cannot be changed.
+
+Storing a Key - from most secure to least secure:
+
+- the key is stored in hardware-backed Android KeyStore
+- all keys are stored on server and are available after strong authentication
+- master key is stored on server and use to encrypt other keys, which are stored in Android SharedPreferences
+- the key is derived each time from a strong user provided passphrase with sufficient length and salt
+- the key is stored in software implementation of Android KeyStore
+- master key is stored in Android Keystore and use to encrypt other keys, which are stored in SharedPreferences
+- all keys are stored in SharedPreferences
+- [not recommended] hardcoded encryption keys in the source code
+- [not recommended] predictable key derivation function based on stable attributes
+- [not recommended] stored generated keys in public places (like `/sdcard/`)
+
+The most secure way of handling key material, is simply never storing it on the device. That can be achieved by using [hardware-backed Android KeyStore](0x05d-Testing-Data-Storage.md#hardware-backed-android-keystore) if device is running Android 7.0 (API level 24) and above with available hardware component (Trusted Execution Environment (TEE) or a Secure Element (SE)). That can be check by using guidelines provided for [the secure implementation of Key Attestation](0x05d-Testing-Data-Storage.md#key-attestation). If hardware component is not available and/or support for Android 6.0 (API level 23) and below is required, then that can be achieved by storing a key on remote server and make a key available after authentication. Limitiation is that app requires always online work.
```

Comment:
The consequences of hosting the keys on the server should be more emphasized to make it clear, as this is a severe downside in some app architectures:

```suggestion
The most secure way of handling key material, is simply never storing it on the device. That can be achieved by using [hardware-backed Android KeyStore](0x05d-Testing-Data-Storage.md#hardware-backed-android-keystore) if device is running Android 7.0 (API level 24) and above with available hardware component (Trusted Execution Environment (TEE) or a Secure Element (SE)). That can be check by using guidelines provided for [the secure implementation of Key Attestation](0x05d-Testing-Data-Storage.md#key-attestation). If hardware component is not available and/or support for Android 6.0 (API level 23) and below is required, then that can be achieved by storing a key on remote server and make a key available after authentication. 

> Please note that if the keys are stored on the server, the app need to be online to decrypt the data. This might be a limitation in some use case of mobile apps and should be carefully thought through as this becomes part of the architecture of the app.
```